### ISSUE TITLE
table missing leading separator errors in unrelated location when missing a "close block" ==== in another part of the document

### ISSUE THREAD

**Issue by janicecmt** [asciidoctor/asciidoctor#2228]
I just turned on a conditional I've had turned off for a while and started getting a series of "table missing leading separator" errors in a completely unrelated location well below the actual cause of the problem (which is unrelated to tables and was only discovered by examining the output and noticing where it started going south).

Basically, I have a conditional that toggles comments meant for reviewers to see and address. I use the following format for these comments so that they stand out in the output:

```
ifdef::review[]
====
Initials: comment
====
endif::[]
```

I just turned it back on after a release and started getting a series of errors:

```
asciidoctor: ERROR: file.adoc: line 236: table missing leading separator, recovering automatically
asciidoctor: ERROR: file.adoc: line 236: table missing leading separator, recovering automatically
asciidoctor: ERROR: file.adoc: line 333: table missing leading separator, recovering automatically
asciidoctor: ERROR: file.adoc: line 236: table missing leading separator, recovering automatically
```

file.adoc is an interior section of table that's reused inside several tables in my document via include statements. By any possible reckoning it does not have 333 lines. Line 236 is a cell entry that starts with |

This table renders properly and gives no errors when I comment out review so the conditionals are not processed.

Since it says it recovered automatically, I took a look at the resulting output and discovered that it did not include the last two parts of my document. All instances of the file above are in the second-to-last part, one of those not rendered in the output. However, the last portion of the last section that is rendered switches the display so that review comments display as regular text and all content that should be outside of a comment is displayed inside a comment block.

I traced the issue back to a missing ==== in the last review comment that displayed normally. Basically, I did the following:

```
ifdef::review[]
====
Initials: comment
endif::[]
```

That's clearly an error and I have no real issue with the resulting rendering. However, the error message is clearly misleading and sent me down the wrong path when chasing down what went wrong. Giving me the wrong file name was almost worse than the wrong type of error, but both were certainly misleading.

Here's a very condensed snippet of code that illustrates my structure (without include files):

```
= Part

content including headers

ifdef::review[]
====
JMK foo
endif::[]

Lots of content here; it is displayed as a separate block

== Header

Lots of content here still part of that same block

ifdef::review[]
====
JMK This is presented as regular text not inside a block
====
endif::[]

Lots of content here part of a new block


= Part break - this is where it stops displaying

Lots of content

== header

Lots of content including subheads

.Table here
|===
| h1
| h2

| c1
| c2
| c3

| c1
| I assume it's complaining about the first table but there's no way to be certain since none of it was rendered; the complaint was around here somewhere, in a row near the bottom of a table with 20+ rows, some of which were themselves commented out by conditionals
| c3

|===

Lots of content

.Another table here
|===
| h1
| h2

| c1
| c2
| c3

| c1
| c2
| c3

|===

More content here
```

**Comment by mojavelinux** [comment]
Thank you for the detailed report.

Unfortunately, line numbers are not always accurate, especially when it involves includes and tables. The only way to really fix this is change how line numbers are tracked. Since we were missing a dedicated issue for this, I added one. See https://github.com/asciidoctor/asciidoctor/issues/2233

**Comment by mojavelinux** [comment]
When Asciidoctor was originally written, it didn't track line numbers at all. We needed to have something in place, so I put a basic solution in place that didn't require a major refactoring of the code. Now we need to take it to the next level and make a more involved change. This feature is also needed now for sync scroll in editors that offer preview.

**Comment by janicecmt** [comment]
Thanks. I expect error messages with line numbers to be slightly off - they are in most cases. However, this one was in the wrong file and many, many lines of content (hundreds if not thousands) away from the actual error. As noted above, the error message was also very confusing because it had nothing to do with what the actual problem was.

Having an accurate line number in the correct file name would help immensely; I would at least know where to look and be able to determine that the error message had nothing to do with the actual issue, but I likely still would have filed an issue over the misleading message even had the other items been correct.

**Comment by mojavelinux** [comment]
I'm confident that this will be solved by the changes coming in 1.5.7, in particular #2701.

As it turns out, the source location was being recorded after the block was read rather than before. This lead to the line number error being compounded with each block.

**Comment by mojavelinux** [comment]
What we're still missing is that the warning message for an unterminated block reports the last line in the reader (usually the file) rather than where the block started. I'm going to file a separate issue for that now that we actually have the right line numbers.

**Comment by mojavelinux** [comment]
Once #2701 is merged, the unterminated block message will report where the block starts. I also corrected a number of other problems with how line numbers were being tracked. I feel confident that the issue reported here has been addressed. If there are still edge cases, we now have a much more solid framework in place for correcting them.

**Comment by mojavelinux** [comment]
resolved by d2b96b232a0e2d88bed056c7df437312b3b8fe00
